src/ray/gcs/gcs_server/gcs_task_manager.cc

```diff
@@ -110,6 +111,30 @@ absl::optional<TaskAttempt> GcsTaskManager::GcsTaskManagerStorage::GetLatestTask
   return latest_task_attempt;
 }
 
+void GcsTaskManager::GcsTaskManagerStorage::MarkTasksFailedOnWorkerDead(
+    const WorkerID &worker_id, const rpc::WorkerTableData &worker_failure_data) {
+  auto task_attempts_itr = worker_to_task_attempt_index_.find(worker_id);
+  if (task_attempts_itr == worker_to_task_attempt_index_.end()) {
+    // No tasks by the worker.
+    return;
+  }
+
+  rpc::RayErrorInfo error_info;
+  error_info.set_error_type(rpc::ErrorType::WORKER_DIED);
+  std::stringstream error_message;
+  error_message << "Worker running the task (" << worker_id.Hex()
+                << ") died with exit_type: " << worker_failure_data.exit_type()
+                << " with error_message: " << worker_failure_data.exit_detail();
+  error_info.set_error_message(error_message.str());
+
+  for (const auto &task_attempt : task_attempts_itr->second) {
+    if (!IsTaskTerminated(task_attempt.first)) {
```

Comment:
Can we just make `IsTaskTerminated` as a part of `MarkTaskAttemptFailed`? And change the API doc to mention it doesn't mark already-terminated tasks to avoid overwriting failure time (feel like there' no reason to decouple this?)

Reply:
good point - i think it always only triggers the non-terminated path.